### ISSUE TITLE
Put config.xml into same folder with documentation data

### ISSUE THREAD

**Issue by adiherzog** [scenarioo/scenarioo#427]
Currently the user needs to configure two folders:
- The folder where config.xml is (this is configured in the context file of the web app)
- The folder where documentation data is (this is configured in config.xml)

We could simplify this and only use one folder for both.

Furthermore we should think about whether the context file is the best place to configure this folder. @xeronimus proposed to use something else like e.g. an environment variable.

The path to the documentation data can currently be changed in the GUI. This should not be possible anymore. This would improve the safety of the application and also prohibit accidental changes of this value.

**Comment by bruderol** [comment]
Maybe some inspiration:
https://wiki.jenkins-ci.org/display/JENKINS/Administering+Jenkins

I think:
1. we should use "~/.scenarioo" as the default location (which is as currently implemented, see ConfigurationDaoImpl, but we should also use this as default for documentation data, which is currently different, see default config.xml).
2. Configuration through a system environment variable seems not the best solution for me, because then it would not be possible to have more than one scenarioo instance on the same server anymore (or not?). The current configuration through initParameters from servlet context, seems good for me. Or what is the disadvantage of this current solution?
3. It is important, that the config.xml is in the same configured folder as the documentaiton, because this makes docker image configuration more easy (but needs to be changed after this change!) and the config.xml is automatically stored together with the other documentation data.

**Comment by adiherzog** [comment]
Good input, I like the Jenkins way. Let's call our root folder `SCENARIOO_HOME` :-)

We could also offer several ways to configure `SCENARIOO_HOME`. The environment variable can be set in a script just before starting a Scenarioo instance. Therefore it would be possible to have several instances with different root folders. Furthermore we can also keep the configuration possibility in the context of the Server. We just have to document the order, in which the configuration is read.

**Comment by bruderol** [comment]
I think, this issue is important to simplify scenarioo usage. 
Therefore I put it to milestone "2.2. - Small Improvements"

And I try to summarize the solution that it seems we agreed upon:

1. config file to be expected in same folder as documentation content.
2. the documentation folder is what can be configured in future as following (take first rule that matches):
- set through webapp context --> use that one
- set through system variable "SCENARIOO_HOME" --> use that one
- otherwise use ~/.scenarioo as default
- not possible to set docu directory through config.xml anymore (breaking change!)
3. consequences to consider:
- on scenarioo update: configuration needs to be migrated somehow (configure docu path, and copy config.xml to docu directory) --> we need to provide migration instructions!
- docker image needs to be updated/changed and documented accordingly

I put this to Improvements, because discussed just now with @mi-we that this would make sense.
Especially for docker image, it would be an important improvement.

Open point: 
Maybe we should only implement that for a special 3.0 release together with other document format changes? Or what you think? Any other ideas for an easy migration path?

@adiherzog Do we want to open a special branch for this? Such that we can start to work on such breaking changes?

**Comment by bruderol** [comment]
I removed "ready to code" ... because of above open point. As soon as we agree on that one, we could start implement that.

**Comment by adiherzog** [comment]
I propose to do this on a feature branch. This allows us to decide in which release we want to include this, which would then make the release a major release because of the breaking change.

**Comment by mi-we** [comment]
Also e2e testing and the demo deployments are affected, since they are using a different config.xml with a specific directory path.

**Comment by mi-we** [comment]
Question about the filename. At the moment it is possible to define a different file name for config.xml via the servlet context. Are there cases, where this functionality is used? If so, should we also offer both ways like with the docu directory (servlet context or env variable)?

The only use case I see is, when you have different config.xml that point to different docu paths, but since the path will not be defined anymore in the config, this use case also doesn't make sense anymore. Maybe we could just strictly define the file name as "config.xml".

@bruderol @adiherzog What do you think?

**Comment by mi-we** [comment]
e2e tests not working yet, working on a solution to safely define the path to always point to the correct directory for the e2e tests.

**Comment by adiherzog** [comment]
I think always using "config.xml" is good, it does not make sense to have this dynamic anymore.

On the build server we have a separate context configuration file for each scenarioo instance, so we can configure the SCENARIOO_ROOT differently for each of them.

@mi-we What's the issue with the e2e tests?

**Comment by bruderol** [comment]
I also see no real need, to keep the config file name dynamic, but for backward compatibility we could leave it, at least for now.

**Comment by bruderol** [comment]
Ähm, wait, since we have to offer a migration guide concerning this change anyway, my last comment concerning "backward compatibility" is kind of nonsense ;-) So, I think we should remove the possibility to use different filenames, but have to consider that in the migration guide (who is writing this?).

**Comment by mi-we** [comment]
@adiherzog I still have to take a look at the e2e tests, they failed last time but maybe it was because of something else. But could be that I broke them because the viewer takes another documentation path now.

OK, I will remove the file name override definitely in this case.

@bruderol I guess it makes sense to at least prepare some wiki page for the migration within this issue. I will take that as a task. (is it possible to write a wiki page and not publish it yet btw?)

**Comment by mi-we** [comment]
OK, I've found the issue with the e2e tests. The problem of course is, that the config.xml for the e2e tests is a special one (src/main/resources/config-for-demo/config.xml) and this has to be copied to the directory of the scenarioo-docu-example. Now, AFAIK this test data is not persistent so we can not just copy the config.xml and it's ok. There are multiple ways now to solve this in my opinion:

1. We replace the default config.xml with the one for the demo. (There are not that many differences so it might be a possible solution)

2. We enhance the test data generation to copy the demo config.xml

3. We don't auto generate the docuExample stuff anymore but make it part of the source (I don't really like this solution)

I'm actually in favor of number 2 at the moment just to make it work. But what  do you think @bruderol @adiherzog ? Should we try to get rid of this "special treatment" for the e2e tests eventually?

**Comment by mi-we** [comment]
Btw, I think, since this is a breaking change anyway, that it makes sense to also not check for the old servlet context property anymore (configurationDirectory) so I will also get rid of that if fine for you.

**Comment by adiherzog** [comment]
Yeah that's fine.

Michael <notifications@github.com> schrieb am Mo., 14. Mär. 2016 16:01:

**Comment by adiherzog** [comment]
I think we just have to adapt the Jenkins build to the new way. Let's discuss / do this on Wednesday.

**Comment by bruderol** [comment]
@mi-we @adiherzog Sorry guys, but I not agree. 

The servlet-context path for the directory should still be supported. Reasons:
1. Because we decided and discussed about this before (please stick to what we discussed before! see my comment about 28 days before and discussion before it)
2. because it is more nicer than through a system variable (don't you think?)
3. because it will make the migration more easier
4. probably most important: because a system variable can only be set once per running tomcat instance and then it would not be possible anymore to deploy more than one scenarioo documentation on one tomcat.

**Comment by bruderol** [comment]
As discussed on #106
We also have to adapt the config UI and make the field for setting the documentation path read only. We should put an information text next to the field that informs the user about our wiki page, where we describe how to configure the documentation path and where to put the config file.

We also have to take care, that the changes to the wiki pages are also done on a branch, because this changes will not yet take effect in the next release. 
@mi-we you can branch the whole wiki for this purpose or simply write new special pages with "(new format version!)" in title for now, and later we can merge this manually, if you not have to change too much pages in the wiki.

**Comment by bruderol** [comment]
@mi-we Regarding config file for e2e-tests:
I think this makes sense, because we have to activate some features (e.g. additional tabs for service calls) for e2e-tests, that should not go in the default configuration. It is (AFAIK) well documented, that for e2e-tests you first need to copy this file accordingly!

**Comment by bruderol** [comment]
@mi-we @adiherzog aha, as discussed with @mi-we today, it is of course okay, to only have one property (not two for backward compatibility anymore) and even to name it differently, e.g. "scenariooDocumentationDirectory", since it is not the configuration directory anymore.

I propose to use "scenariooDataDirectory" or something similar as a name for the property, that explains a bit more, what it is for (not just "SCENARIOO_HOME"). Maybe also for the environment variable? What do you think?

**Comment by mi-we** [comment]
@bruderol As discussed with you yesterday I agree on also changing the environment variable as it is not really the "home" directory of Scenarioo. I try to summarise the last few discussions to the following points:

- Servlet context property will be renamed to `scenariooDocumentationDirectory`
- Fallback to old servlet context property `configurationDirectory` will be removed
- Documentation path is shown as read-only text under "Manage" (already done) and tooltip or dialog will provide link to the Wiki page on how to set it
- The wiki will contain a migration guide (either with specially marked pages or on a separate branch of the wiki)

Hope I didn't miss something.

**Comment by mi-we** [comment]
I created a draft now for the migration guide. You can find it here:

https://github.com/scenarioo/scenarioo/wiki/Migration-guide-from-2.x-to-3.x-%28new-format-version!%29

Unfortunately Github does not seem to support multiple branches and pull requests for the Wiki, so for now I created the pages with a remark that they belong to the next bigger Scenarioo release.

**Comment by bruderol** [comment]
@mi-we Ich hätte noch ein paar Verbesserungs-Vorschläge für das Wiki, lass uns das mal bei Gelegenheit besprechen. Ich denke auch, dass sich das sichelrich branchen lässt, auch im Hinblick auf die Anpassungen für Format 3.0, wäre das dann wichtig.

**Comment by bruderol** [comment]
@mi-we I reviewed a little bit your code changes, only roughly, and I liked a lot what I saw! Good work! Thank you for your contributions! :+1:

**Comment by mi-we** [comment]
@bruderol Thanks for taking a look at it! Regarding the wiki I just found the following StackOverflow question:

http://stackoverflow.com/questions/14783486/github-wiki-select-branch-dispalyed-on-github

So you're right, it is possible to branch but only the master branch is published on Github. I will move the two new pages to a separate branch in this case.

**Comment by bruderol** [comment]
@mi-we perfect. Just let me know, when you want to do a review of the wiki pages. I think we should do that face-2-face. Thank you!

**Comment by mi-we** [comment]
I removed the two new pages from the wiki master branch and created a separate branch for the next Scenarioo major release. The wiki branch is called: Scenarioo3.0

**Comment by adiherzog** [comment]
Shouldn't we look for a solution where we can use the code repository for
the documentation? Then it would automatically be versioned with the code.

With the current approach we always have to make sure we merge the wiki
repo whenever we merge the code repo.

Michael <notifications@github.com> schrieb am Mo., 21. Mär. 2016 16:34:

**Comment by bruderol** [comment]
Yes, exactly, that is also my long term goal!

**Comment by bruderol** [comment]
@mi-we @adiherzog 

Just stumbled over this when trying to setup my IntelliJ for scenarioo development:

We should simply include the default config.xml for the demo in the generation of the example docu data. I mean that running this tests should also copy the default configuration there.

This will help us with two things:
1. We ensure that on local development, the right configuration for e2e-tests is allready there and will be reset on every rerun of the example-data-generator
2. Also for CI this will make it more easy: just copy the generated data folder in the right place (including the config.xml)

This is quite easy now, that the config.xml has become independent of any ugly path :+1: 

I could help you with this part, if you need help for this. Just let me know. when do you plan to merge this to 'develop'?

**Comment by bruderol** [comment]
please see the TODO comment in my above commit. To be resolved, when merge has been done to develop. My new IntelliJ Dev Env setup, seems to work so far, and is allready prepared for the changes of this issue :smile:

Actually this issue makes it a lot easier to configure scenarioo for development, if we do it right (see other comment above).

**Comment by bruderol** [comment]
A changed the documentation about how to setup a scenarioo tomcat server (because I am doing so in my current project). See here:
https://github.com/scenarioo/scenarioo/wiki/Scenarioo-Viewer-Web-Application-Setup

The new configuration instruction allready informs the users that this configuration will change in future scenarioo versions and that we recommend to use same directory for config dir as well as for documentation path to be allready prepared for this changes in future scenarioo versions.

**Comment by bruderol** [comment]
@mi-we @adiherzog I think it would be important to finish this on the coding days. Also the docker image needs some adjustments and we should really start now to move all our documentation for version 3.0 from the Wiki into Md-Files directly checked in into our sources. I started this in another education project and it is just a fantastic approach with many many many benefits!

**Comment by adiherzog** [comment]
@mi-we @bruderol As decided yesterday this change should not be merged into the develop branch yet as it will be released with the format 3.0 later this year.

**Comment by bruderol** [comment]
yes, I will have a look at this together with @mi-we

**Comment by bruderol** [comment]
I renamed milestones 3.0 to 2.3 and moved this issue to 3.0 - Take care to move anything out of 2.3 if it is a breaking change

**Comment by mi-we** [comment]
Closing this now, functionality is in productive version since 4.x now
